File: src/main/java/fi/quanfoxes/parser/nodes/TypeNode.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Contextable;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Parser;
import fi.quanfoxes.parser.Type;

import java.util.ArrayList;
import java.util.List;

public class TypeNode extends Node implements Contextable {
    private Type type;
    private List<Token> body;

    public TypeNode(Type type) {
        this(type, new ArrayList<>());
    }

    public TypeNode(Type type, List<Token> body) {
        this.type = type;
        this.body = body;
    }

    public void parse() throws Exception {
        Parser.parse(this, type, body, Parser.MEMBERS);
        body.clear();
    }

    public Type getType() {
        return type;
    }

    @Override
    public Context getContext() {
        return type;
    }

    @Override
    public NodeType getNodeType() {
        return NodeType.TYPE_NODE;
    }
}

File: Zigzag/Parser/Nodes/TypeNode.cs
using System;
using System.Collections.Generic;

public class TypeNode : Node, IType, IContext
{
	public Type Type { get; private set; }
	public bool IsDefinition { get; private set; } = false;

	private List<Token> Body { get; set; }

	public TypeNode(Type type)
	{
		Type = type;
		Body = new List<Token>();
	}

	public TypeNode(Type type, List<Token> body)
	{
		Type = type;
		Body = body;
		IsDefinition = true;
	}

	public void Parse()
	{
		Parser.Parse(this, Type, Body);
		Body.Clear();

		// Find all expressions which represent type initialization
		var expressions = FindChildren(n => !n.Is(NodeType.FUNCTION_DEFINITION_NODE) && !n.Is(NodeType.VARIABLE_NODE));

		// Pack all expressions under an initialization node
		var initialization = new Node();
		expressions.ForEach(e => initialization.Add(e));

		// Save the initialization for other constructors
		Type.Initialization = initialization;
	}

	public new Type? GetType()
	{
		return Type;
	}

	public override NodeType GetNodeType()
	{
		return NodeType.TYPE_NODE;
	}

	public override bool Equals(object? obj)
	{
		return obj is TypeNode node &&
				base.Equals(obj) &&
				EqualityComparer<Type>.Default.Equals(Type, node.Type);
	}

	public override int GetHashCode()
	{
		var hash = new HashCode();
		hash.Add(base.GetHashCode());
		hash.Add(Type.Name);
		return hash.ToHashCode();
	}

	public void SetContext(Context context)
	{
		throw new InvalidOperationException("Replacing context of a type node is not allowed");
	}

   public Context GetContext()
   {
		return Type;
   }
}
